Make aes_hw_ctr32_encrypt_blocks handle len=0 correctly

nebeid opened the pull request:
### Issues:
Resolves #CryptoAlg-2498

### Description of changes: 
`aes_hw_ctr32_encrypt_blocks` encrypts (and writes to the output) 2 blocks when the input length is 0 blocks in the case of AArch64 and 1 block in the case of x86_64 and x86.
The function is guarded wherever it's called by checks that `len != 0`.
This change fixes this behaviour without taxing the performance.

### Testing:
Tested the performance on Graviton3 and on Mac x86_64.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

andrewhop reviewed (commented):
Can you add a test that calls `aes_hw_ctr32_encrypt_blocks` with a length of zero and verifies the input/output buffers aren't modified. Does address sanitizer find this original issue without this change?

nebeid commented:
> Can you add a test that calls `aes_hw_ctr32_encrypt_blocks` with a length of zero and verifies the input/output buffers aren't modified. Does address sanitizer find this original issue without this change?

Yeah, I had added the test here https://github.com/aws/aws-lc/pull/1690/commits/69b858ac385a0a73c8f92a2e67d672cb0153425d with printouts to verify for myself. I reverted that commit which included the fix for x86 (due to performance regression). I can add back the test `#ifdef`ed for AArch64.

nebeid commented:
amazonlinux2023_gcc11x_aarch_valgrind complained
```
[ RUN      ] AESTest.ABI
==3382== Conditional jump or move depends on uninitialised value(s)
==3382==    at 0x504B60: bssl::internal::operator==(bssl::Span<unsigned char const>, bssl::Span<unsigned char const>) (in /codebuild/output/src1610932754/src/github.com/aws/aws-lc/test_build_dir/crypto/crypto_test)
==3382==    by 0x5039EB: operator==(Bytes const&, Bytes const&) (in /codebuild/output/src1610932754/src/github.com/aws/aws-lc/test_build_dir/crypto/crypto_test)
==3382==    by 0x510C5F: testing::AssertionResult testing::internal::CmpHelperEQ<Bytes, Bytes>(char const*, char const*, Bytes const&, Bytes const&) (in /codebuild/output/src1610932754/src/github.com/aws/aws-lc/test_build_dir/crypto/crypto_test)
==3382==    by 0x504E83: testing::AssertionResult testing::internal::EqHelper::Compare<Bytes, Bytes, (void*)0>(char const*, char const*, Bytes const&, Bytes const&) (in /codebuild/output/src1610932754/src/github.com/aws/aws-lc/test_build_dir/crypto/crypto_test)
==3382==    by 0x697383: AESTest_ABI_Test::TestBody() (in /codebuild/output/src1610932754/src/github.com/aws/aws-lc/test_build_dir/crypto/crypto_test)
==3382==    by 0x8F25BF: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (in /codebuild/output/src1610932754/src/github.com/aws/aws-lc/test_build_dir/crypto/crypto_test)
==3382==    by 0x8ED0D3: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (in /codebuild/output/src1610932754/src/github.com/aws/aws-lc/test_build_dir/crypto/crypto_test)
==3382==    by 0x8D1ACB: testing::Test::Run() (in /codebuild/output/src1610932754/src/github.com/aws/aws-lc/test_build_dir/crypto/crypto_test)
==3382==    by 0x8D21C3: testing::TestInfo::Run() (in /codebuild/output/src1610932754/src/github.com/aws/aws-lc/test_build_dir/crypto/crypto_test)
==3382==    by 0x8D28A7: testing::TestSuite::Run() (in /codebuild/output/src1610932754/src/github.com/aws/aws-lc/test_build_dir/crypto/crypto_test)
==3382==    by 0x8DB943: testing::internal::UnitTestImpl::RunAllTests() (in /codebuild/output/src1610932754/src/github.com/aws/aws-lc/test_build_dir/crypto/crypto_test)
==3382==    by 0x8F3593: bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) (in /codebuild/output/src1610932754/src/github.com/aws/aws-lc/test_build_dir/crypto/crypto_test)
==3382==  Uninitialised value was created by a stack allocation
==3382==    at 0x696764: AESTest_ABI_Test::TestBody() (in /codebuild/output/src1610932754/src/github.com/aws/aws-lc/test_build_dir/crypto/crypto_test)
==3382== 
[       OK ] AESTest.ABI (54 ms)
```
So I'll try comparing `std::string` instead of `Bytes`.

andrewhop commented on crypto/fipsmodule/aes/aes_test.cc:
```diff
@@ -479,6 +479,17 @@ TEST(AESTest, ABI) {
+          for (size_t i = 0; i < 64; i++) {
+            buf[i] = i;
+          }
+          Bytes buf_before = Bytes(buf,64);
```
Bytes/span doesn't create a copy of the data, they reference the original memory. So this test passes even if aes_hw_encrypt touches buf. For example this passes:
```
TEST(BytesTest, copyOrClone) {
  uint8_t buf[64] = {0};
  Bytes buf_before = Bytes(buf,64);
  buf[0] = 1;
  ASSERT_EQ(buf_before, Bytes(buf, 64));
}
```

nebeid replied:
Thanks, Andrew. I brought back the string comparison I had as I was testing in https://github.com/aws/aws-lc/pull/1690/commits/110ab16bdea385c42881473e71f677043b75ae59.

dkostic commented on crypto/fipsmodule/aes/asm/aesni-x86_64.pl:
```diff
@@ -1234,6 +1234,7 @@ sub aesni_generate8 {
 
 .align	16
 .Lctr32_bulk:
+	jb .Lctr32_epilogue			# if $len < 1, go to done
```
shouldn't this be right after `cmp` in line 1216?
